### PR TITLE
pre-commit: Add ruff to replace bandit, flake8, isort, and pyupgrade

### PR DESCRIPTION
**Pull request by cclauss**
Use https://beta.ruff.rs to replace several Python linting tools in pre-commit with a unified tool written in Rust for speed.

Migrate from `setup.cfg` to `pyproject.toml` in alignment with [PEP-621](https://peps.python.org/pep-0621).

Other minor fixes to comply with Ruff rules.

Other Python linters can easily be added via simple edits to the `tool.ruff.select` variable in `pyproject.toml`:
% `ruff linter`
```
   F Pyflakes
 E/W pycodestyle
 C90 mccabe
   I isort
   N pep8-naming
   D pydocstyle
  UP pyupgrade
 YTT flake8-2020
 ANN flake8-annotations
   S flake8-bandit
 BLE flake8-blind-except
 FBT flake8-boolean-trap
   B flake8-bugbear
   A flake8-builtins
 COM flake8-commas
  C4 flake8-comprehensions
 DTZ flake8-datetimez
 T10 flake8-debugger
  DJ flake8-django
  EM flake8-errmsg
 EXE flake8-executable
 ISC flake8-implicit-str-concat
 ICN flake8-import-conventions
   G flake8-logging-format
 INP flake8-no-pep420
 PIE flake8-pie
 T20 flake8-print
 PYI flake8-pyi
  PT flake8-pytest-style
   Q flake8-quotes
 RSE flake8-raise
 RET flake8-return
 SLF flake8-self
 SIM flake8-simplify
 TID flake8-tidy-imports
 TCH flake8-type-checking
 ARG flake8-unused-arguments
 PTH flake8-use-pathlib
 ERA eradicate
  PD pandas-vet
 PGH pygrep-hooks
  PL Pylint
 TRY tryceratops
 NPY NumPy-specific rules
 RUF Ruff-specific rules
```

### CODE REVIEW

**Comment by cclauss** [comment]
The playwright timeout issues seem unrelated to these changes.

**Review by hoodmane — APPROVED** [review]
Thanks @cclauss.

**Comment by marimeireles** [comment]
@hoodmane should I make the changes you proposed and merge this one?

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -20,51 +20,31 @@ repos:
...
--   repo: https://github.com/PyCQA/bandit
-    rev: 1.7.4
+-   repo: https://github.com/charliermarsh/ruff-pre-commit
+    rev: v0.0.247
     hooks:
-    -   id: bandit
-        args:
-        -   --skip=B101,B201
+    -   id: ruff
+        # args: [--fix, --exit-non-zero-on-fix]
```
Comment (cclauss) [inline .pre-commit-config.yaml:27]:
```suggestion
```

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -20,51 +20,31 @@ repos:
...
-    -   id: isort
-        name: isort (python)
-        args: [--profile, black]
+    -   id: codespell  # See 'pyproject.toml' for args
+        additional_dependencies:
+        -   tomli
 
 -   repo: https://github.com/macisamuele/language-formatters-pre-commit-hooks
-    rev: v2.6.0
+    rev: v2.7.0
```
Comment (cclauss) [inline .pre-commit-config.yaml:42]:
```suggestion
    rev: v2.6.0
```

**Review by hoodmane — APPROVED** [review]
Thanks @cclauss!

**Review by FabioRosado — APPROVED** [review]
thanks 😄

**Pull request merged**